File: languages/scala/src/start.scala
import java.io.{FileInputStream, InputStream}
import java.util.Properties

import model.exception.InvalidParameterException
import model.{BenchmarkOutput, TableAnalysis, TimeFormat, UserInfo}
import provider._

object main {

  def main(args: Array[String]): Unit = {

    val configFile = "config.properties"
    val properties = loadProperties(configFile)

    val input = properties.getProperty("INPUT_FILENAME")
    val output = properties.getProperty("OUTPUT_FILENAME")

    val benchmark: BenchmarkOutput = new BenchmarkMeasure

    val summaryAnalysis: TableAnalysis[Array[Double]] = new SummaryAnalysis
    val bubbleSortAnalysis: TableAnalysis[List[UserInfo]] = new BubbleSortAnalysis
    val quickSortAnalysis: TableAnalysis[List[UserInfo]] = new QuickSortAnalysis

    //==================================================
    // Leitura dos dados
    benchmark.start("Read")
    val table = new TableReader(input)
    val list = table.read()
    benchmark.end("Read")
    //==================================================
    // Analise dos dados (Summary)
    benchmark.start("SummaryAnalyse")
    val summaryValue: Array[Double] = summaryAnalysis.analysis(list)
    benchmark.end("SummaryAnalyse")
    //==================================================
    // Analise dos dados (Bubble)
    benchmark.start("BubbleAnalyse")
    val bubbleValue: List[UserInfo] = bubbleSortAnalysis.analysis(list)
    benchmark.end("BubbleAnalyse")
    //==================================================
    // Analise dos dados (Quick)
    benchmark.start("QuickAnalyse")
    val quickValue: List[UserInfo] = quickSortAnalysis.analysis(list)
    benchmark.end("QuickAnalyse")
    //==================================================

    benchmark.export(output, TimeFormat.MILISSEGUNDOS)
  }

  def loadProperties(configFile: String): Properties = {
    if (configFile == null) throw new InvalidParameterException("'configFile' é null")

    val input = new FileInputStream (configFile)
    val properties = new Properties
    properties.load (input)

    return properties
  }

}

File: languages/java/src/Start.java
import model.BenchmarkOutput;
import model.TableAnalysis;
import model.TimeFormat;
import model.exception.DataReaderException;
import model.exception.InvalidParameterException;
import provider.*;
import model.UserInfo;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Properties;

public class Start {

    public static void main(String args[]) throws IOException {

        if(args.length == 0) throw new InvalidParameterException("Parametros iniciais nao encontrados");

        String configFile = args[0];
        Properties properties = loadProperties(configFile);

        String input = properties.getProperty("INPUT_FILENAME");
        String output = properties.getProperty("OUTPUT_FILENAME");

        BenchmarkOutput benchmark = new BenchmarkMeasure();

        TableAnalysis summaryAnalysis = new SummaryAnalysis();
        TableAnalysis bubbleSortAnalysis = new BubbleSortAnalysis();
        TableAnalysis quickSortAnalysis = new QuickSortAnalysis();
        TableAnalysis languageSortAnalysis = new LanguageSortAnalysis();

        //==================================================
        // Leitura dos dados
        benchmark.start("Read");
        TableReader tableReader = new TableReader(input);
        List<UserInfo> list = tableReader.read();
        benchmark.end("Read");
        //==================================================
        // Analise dos dados (Summary)
        benchmark.start("SummaryAnalyse");
        double[] summary = (double[]) summaryAnalysis.analysis(list);
        benchmark.end("SummaryAnalyse");
        //==================================================
        // Analise dos dados (Bubble)
        benchmark.start("BubbleAnalyse");
        List<UserInfo> bubble = (List<UserInfo>) bubbleSortAnalysis.analysis(list);
        benchmark.end("BubbleAnalyse");
        //==================================================
        // Analise dos dados (Quick)
        benchmark.start("QuickAnalyse");
        List<UserInfo> quick = (List<UserInfo>) quickSortAnalysis.analysis(list);
        benchmark.end("QuickAnalyse");
        //==================================================
        // Analise dos dados (Language)
        benchmark.start("LanguageAnalyse");
        List<UserInfo> lang = (List<UserInfo>) languageSortAnalysis.analysis(list);
        benchmark.end("LanguageAnalyse");
        //==================================================

        benchmark.export(output, TimeFormat.MILLISEGUNDOS);
    }

    public static Properties loadProperties(String configFile) throws IOException {
        if(configFile == null) throw new InvalidParameterException("'configFile' é null");

        try {
            InputStream input = new FileInputStream(configFile);
            Properties properties = new Properties();
            properties.load(input);

            return properties;

        } catch(IOException e) {
            throw new DataReaderException("Erro ao ler arquivo:" + configFile);
        }
    }
}
